### PR TITLE
Document configuration migration to 2.0 steps

### LINKED ISSUE #4277: Document configuration migration to 2.0
Per the request in https://github.com/aspnet/Configuration/issues/728, add a new section to the [main migration doc](https://docs.microsoft.com/aspnet/core/migration/1x-to-2x/).

### PR DESCRIPTION
**Pull request by isaac2004**
Fixes #4277 

Added section to Core 1.x to 2.0 Migration Doc to show example of adding additional configuration providers since the boilerplate configuration is done behind the scenes in 2.0

[Internal Review Page](https://review.docs.microsoft.com/aspnet/core/migration/1x-to-2x/index?branch=pr-en-us-4453#add-configuration-providers)

### CODE REVIEW

**Comment by isaac2004** [comment]
@scottaddie please review

File: aspnetcore/migration/1x-to-2x/index.md [hunk 1]
```diff
@@ -98,6 +98,20 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
 ```
 Unable to create an object of type '<Context>'. Add an implementation of 'IDesignTimeDbContextFactory<Context>' to the project, or see https://go.microsoft.com/fwlink/?linkid=851728 for additional patterns supported at design time.
 ```
+## Adding Additional Configurations
+
+In 2.0 projects, the boilerplate configuration code is done behind the scenes. By default, Environment Variables and Application Settings are loaded into startup so all that is required in *Startup.cs* is initializing `IConfiguration` with the injected instance. To add additional providers, or change the existing providers, utilize `ConfigureAppConfiguration` on `IWebHostBuilder` in *Program.cs*. The setup is similar to Core 1.x
```
Comment (scottaddie) [inline aspnetcore/migration/1x-to-2x/index.md]: ~~In 2.0 projects, the boilerplate configuration code is done behind the scenes.~~
In 2.0 projects, the boilerplate configuration code runs behind-the-scenes.

- Environment Variables --> environment variables
- Application Settings --> app settings
- Core 1.x --> 1.x:

File: aspnetcore/migration/1x-to-2x/index.md [hunk 2]
```diff
@@ -98,6 +98,20 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+
+````
```
Comment (scottaddie) [inline aspnetcore/migration/1x-to-2x/index.md]: ```` --> ```csharp

File: aspnetcore/migration/1x-to-2x/index.md [hunk 3]
```diff
@@ -98,6 +98,20 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+ var host = new HostBuilder()
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.AddEnvironmentVariables();
+                    config.AddJsonFile("appsettings.json", optional: true);
+                    config.AddCommandLine(args);
+                })
+````
```
Comment (scottaddie) [inline aspnetcore/migration/1x-to-2x/index.md]: ```` --> ```

File: aspnetcore/migration/1x-to-2x/index.md [hunk 4]
```diff
@@ -98,6 +98,20 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+In 2.0 projects, the boilerplate configuration code is done behind the scenes. By default, Environment Variables and Application Settings are loaded into startup so all that is required in *Startup.cs* is initializing `IConfiguration` with the injected instance. To add additional providers, or change the existing providers, utilize `ConfigureAppConfiguration` on `IWebHostBuilder` in *Program.cs*. The setup is similar to Core 1.x
+
+````
+ var host = new HostBuilder()
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.AddEnvironmentVariables();
+                    config.AddJsonFile("appsettings.json", optional: true);
+                    config.AddCommandLine(args);
+                })
```
Comment (scottaddie) [inline aspnetcore/migration/1x-to-2x/index.md]: Add missing semicolon at the end of this line

File: aspnetcore/migration/1x-to-2x/index.md [hunk 5]
```diff
@@ -98,6 +98,20 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
 ```
 Unable to create an object of type '<Context>'. Add an implementation of 'IDesignTimeDbContextFactory<Context>' to the project, or see https://go.microsoft.com/fwlink/?linkid=851728 for additional patterns supported at design time.
 ```
+## Adding Additional Configurations
+
+In 2.0 projects, the boilerplate configuration code is done behind the scenes. By default, Environment Variables and Application Settings are loaded into startup so all that is required in *Startup.cs* is initializing `IConfiguration` with the injected instance. To add additional providers, or change the existing providers, utilize `ConfigureAppConfiguration` on `IWebHostBuilder` in *Program.cs*. The setup is similar to Core 1.x
+
+````
+ var host = new HostBuilder()
```
Comment (scottaddie) [inline aspnetcore/migration/1x-to-2x/index.md]: HostBuilder --> WebHostBuilder

**Comment by scottaddie** [comment]
@davidfowl Please review.

**Comment by davidfowl** [comment]
We should call out what gets added by default and show a before and after (1.x template vs 2.x template)

**Comment by isaac2004** [comment]
@davidfowl I added information regarding 1.x, let me know if anything needs tweaks

File: aspnetcore/migration/1x-to-2x/index.md [hunk 6]
```diff
@@ -99,6 +99,65 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
 
+<a name="add-modify-configuration"></a>
+
+## Add or modify configuration providers
+
+In 1.x projects, adding configuration providers to an application was done during the middleware process by creating an instance of `ConfigurationBuilder`, loading applicable providers (environment variables, app settings, etc), and initializing a a member of `IConfigurationRoot`.
+
+```csharp
+public Startup(IHostingEnvironment env)
+    {
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: nit: fix the formatting

File: aspnetcore/migration/1x-to-2x/index.md [hunk 7]
```diff
@@ -99,6 +99,65 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
+                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
+                .AddEnvironmentVariables();
+        Configuration = builder.Build();
+    }
+
+public IConfigurationRoot Configuration { get; }        
+
+```
+The above example loads the `Configuration` member with environment variables supplied in *launchSettings.json* and configuration settings from *appsettings.json* as well as any appsettings file that matches the `EnvironmentName` property provided in `IHostingEnvironment`. The location of these files would be at the same path as `Startup.cs`
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: I'd leave out the `launchSettings.json` comment. That's a visual studio, developer artifact only.

File: aspnetcore/migration/1x-to-2x/index.md [hunk 8]
```diff
@@ -99,6 +99,65 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+        .ConfigureAppConfiguration((hostContext, config) =>
+        {
+            config.AddEnvironmentVariables();
+            config.AddJsonFile("appsettings.json", optional: true);
+            config.AddCommandLine(args);
+        })
+        .Build();
+```
+
+The configuration used by the `CreateDefaultBuilder` method in the preceding code snippet can be seen [here](https://github.com/aspnet/MetaPackages/blob/396f413b389f5983e413be465f917cbfdbda26c9/src/Microsoft.AspNetCore/WebHost.cs#L152).
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: Don't point to the sha, point to the release:

https://github.com/aspnet/MetaPackages/blob/rel/2.0.0/src/Microsoft.AspNetCore/WebHost.cs#L152

File: aspnetcore/migration/1x-to-2x/index.md [hunk 9]
```diff
@@ -99,6 +99,65 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+{
+    public Startup(IConfiguration configuration)
+    {
+        Configuration = configuration;
+    }
+
+    public IConfiguration Configuration { get; }
+```
+
+To add or modify configuration providers, invoke the `IWebHostBuilder.ConfigureAppConfiguration` method in *Program.cs*:
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: - You can't modify, you can only add more providers.
- We should also show an example of clearing all previous providers and adding new ones.

Comment (isaac2004) [reply aspnetcore/migration/1x-to-2x/index.md]: @davidfowl is there a more elegant way to resetting providers than just recreating `IConfigurationRoot` in *Startup.cs* constructor? Not seeing any extension on `IWebHostBuilder` to clear or remove configurations.

File: aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs [hunk 10]
```diff
@@ -12,6 +12,7 @@ namespace AspNetCoreDotNetFx2._0App
 {
     public class Program
     {
+        #region snippet_ProgramMainConfigProviders
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs]: Regions are the devil, don't use them 😄

File: aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs [hunk 11]
```diff
@@ -20,6 +21,13 @@ public static void Main(string[] args)
         public static IWebHost BuildWebHost(string[] args) =>
             WebHost.CreateDefaultBuilder(args)
                 .UseStartup<Startup>()
+                .ConfigureAppConfiguration((hostContext, config) =>
+                {
+                    config.AddEnvironmentVariables();
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs]: These aren't good ones to show. We should add things that aren't added by default:

```c#
config.AddJsonFile("myconfig.json");
```

File: aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs (hunk 10 above)
Comment (scottaddie) [reply aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs]: I agree that they're the devil; however, these regions are used specifically to make importing of code snippets into the docs easier to maintain. Indicating line numbers has proven very brittle, especially as code samples change.

File: aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs (hunk 11 above)
Comment (isaac2004) [reply aspnetcore/migration/1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Program.cs]: @davidfowl done... can you advise on resetting the providers?

**Comment by scottaddie** [comment]
@davidfowl Please review, and see the question from @isaac2004 above.

**Comment by davidfowl** [comment]
Did we get rid of the example that clears the previous providers?

**Comment by isaac2004** [comment]
@davidfowl I asked earlier if there is an elegant way to do that. Looking at the code in Hosting there doesn't seem to be an extension.

**Comment by davidfowl** [comment]
@isaac2004 `config.Providers.Clear();`

**Comment by isaac2004** [comment]
@davidfowl so doing that does not seem to clear the Providers. Here is a sample Startup

````
    public Startup(IConfiguration config, IHostingEnvironment env)
    {

      var configRoot = (IConfigurationRoot)config;
      configRoot.Providers.ToList().Clear();

      var builder = new ConfigurationBuilder()
           .SetBasePath(env.ContentRootPath)
           .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
           .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
           .AddEnvironmentVariables();
      Configuration = builder.Build();
    }

````

If I run that, the Providers is not modified at all. Am I missing something here?

**Comment by davidfowl** [comment]
That’s not right. The clear happens in ConfigureAppConfiguration. You clear the existing providers on the builder and add new ones

**Comment by isaac2004** [comment]
But you don't have a public member of providers until after you Build it. You can't do it in ConfigureAppConfiguration it is still a Builder

**Comment by isaac2004** [comment]
@davidfowl It looks like Providers is actually Sources on the Builder. I added a sample of Clear, please review

**Comment by scottaddie** [comment]
@isaac2004 Disclaimer: If that's wrong, you can blame me! 😄

File: aspnetcore/migration/1x-to-2x/index.md [hunk 12]
```diff
@@ -99,6 +99,27 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
 Unable to create an object of type '<Context>'. Add an implementation of 'IDesignTimeDbContextFactory<Context>' to the project, or see https://go.microsoft.com/fwlink/?linkid=851728 for additional patterns supported at design time.
 ```
 
+<a name="add-modify-configuration"></a>
+
+## Add configuration providers
+In 1.x projects, adding configuration providers to an app was accomplished during the middleware registration process. The steps involved creating an instance of `ConfigurationBuilder`, loading applicable providers (environment variables, app settings, etc.), and initializing a member of `IConfigurationRoot`.
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: >  adding configuration providers to an app was accomplished during the middleware registration process

adding configuration providers to an app was accomplished via the Startup constructor.

File: aspnetcore/migration/1x-to-2x/index.md [hunk 13]
```diff
@@ -99,6 +99,27 @@ The adoption of this new 2.0 pattern is highly recommended and is required for p
...
+
+[!code-csharp[Main](../1x-to-2x/samples/AspNetCoreDotNetCore1App/AspNetCoreDotNetCore1App/Startup.cs?name=snippet_1xStartup)]
+
+The preceding example loads the `Configuration` member with configuration settings from *appsettings.json* as well as any *appsettings.\<EnvironmentName\>.json* file matching the `IHostingEnvironment.EnvironmentName` property. The location of these files is at the same path as *Startup.cs*.
+
+In 2.0 projects, the boilerplate configuration code inherent to 1.x projects runs behind-the-scenes. For example, environment variables and app settings are loaded at startup. The equivalent *Startup.cs* code is reduced to `IConfiguration` initialization with the injected instance:
+
+[!code-csharp[Main](../1x-to-2x/samples/AspNetCoreDotNetFx2.0App/AspNetCoreDotNetFx2.0App/Startup.cs?name=snippet_2xStartup)]
+
+To remove all providers, invoke the `Clear` method on the `IConfigurationBuilder.Sources` property inside of `ConfigureAppConfiguration`. To add providers back, utilize the `ConfigureAppConfiguration` method in *Program.cs*:
```
Comment (davidfowl) [inline aspnetcore/migration/1x-to-2x/index.md]: > To remove all providers

To remove the default providers added by `WebHostBuilder.CreateDefaultBuilder`.

**Comment by scottaddie** [comment]
@isaac2004 Thank you for the awesome work on this one! We really appreciate the time you put into it and the positive impact it will ultimately make on the community. Your changes will go live within the next 24 hours.

@davidfowl Thank you to you as well for the thorough reviews!

**Comment by davidfowl** [comment]
Thanks for adding this guys!

**Pull request merged**